### PR TITLE
swarm: fix recursive resolving of DNS multiaddrs

### PR DESCRIPTION
**Pull request by marten-seemann**

### CODE REVIEW

**Comment by marten-seemann** [comment]
The code works, but I should add a test case before we merge this.

**Review by sukunrt — COMMENTED** [review]
This change looks fine to me. But should https://github.com/libp2p/go-libp2p/blob/master/p2p/transport/websocket/websocket.go#L139 ignore `dnsaddr` addresses since the actual host will be obtained after resolving the `dnsaddr`

**Comment by dennis-tra** [comment]
Just smoked-tested it with [`boomo`](https://github.com/plprobelab/boomo) and it works:

```
boomo_up{otel_scope_name="github.com/plprobelab/boomo",otel_scope_version="",peer="QmNnooDu7bfjPFoTZYxMNLWUQJyrVwtbZg5gBMjTezGAJN",transport="ws"} 1
boomo_up{otel_scope_name="github.com/plprobelab/boomo",otel_scope_version="",peer="QmQCU2EcMqAqQPR2i9bChDtGNJchTbq5TbXJJ16u19uLTa",transport="ws"} 0
boomo_up{otel_scope_name="github.com/plprobelab/boomo",otel_scope_version="",peer="QmaCpDMGvV2BGHeYERUEnRQAwe3N8SzbUtfsmvsqQLuvuJ",transport="ws"} 0
boomo_up{otel_scope_name="github.com/plprobelab/boomo",otel_scope_version="",peer="QmbLHAnMoJPWSCR5Zhtx6BHJX9KiKNN6tpvbUcqanj75Nb",transport="ws"} 1
boomo_up{otel_scope_name="github.com/plprobelab/boomo",otel_scope_version="",peer="QmcZf59bWwK5XFi76CZX8cbJ4BhTzzA3gU1ZjYZcYW3dwt",transport="ws"} 1
```

(see the `1` at the end of the prometheus metric - at least for 3/5 bootstrappers). I guess `QmQCU2EcMqAqQPR2i9bChDtGNJchTbq5TbXJJ16u19uLTa` isn't working because rust doesn't support ws? `QmaCpDMGvV2BGHeYERUEnRQAwe3N8SzbUtfsmvsqQLuvuJ` is only configured with tcp.

**Comment by marten-seemann** [comment]
I added a test case, and linked the original issue.

File: p2p/net/swarm/swarm_dial_test.go [hunk 1]
```diff
@@ -263,6 +250,35 @@ func TestAddrResolutionRecursive(t *testing.T) {
...
+func TestAddrResolutionRecursiveTransportSpecific(t *testing.T) {
+	p := test.RandPeerIDFatal(t)
+
+	backend := &madns.MockResolver{
+		IP: map[string][]net.IPAddr{
+			"sub.example.com": {net.IPAddr{IP: net.IPv4(1, 2, 3, 4)}},
+		},
+		TXT: map[string][]string{
+			"_dnsaddr.example.com":             {"dnsaddr=/dns4/sub.example.com/tcp/443/wss/p2p/" + p.String()},
+			"_dnsaddr.am6.bootstrap.libp2p.io": {"dnsaddr=/ip4/1.2.3.4/tcp/443/wss"},
```
Comment (sukunrt) [inline p2p/net/swarm/swarm_dial_test.go:263]: Do you need this line? 
>			"_dnsaddr.am6.bootstrap.libp2p.io": {"dnsaddr=/ip4/1.2.3.4/tcp/443/wss"},

**Review by sukunrt — APPROVED** [review]

**Pull request merged**